server/auth/types/multiple/multi_auth.ts

```diff
@@ -0,0 +1,192 @@
+import { BasicAuthentication, OpenIdAuthentication, SamlAuthentication } from '../../types';
+
+export class MultipleAuthentication extends AuthenticationType {
+  private authTypes: string | string[];
+  private BasicAuth!: AuthenticationType;
+  private SamlAuth!: AuthenticationType;
+  private OidcAuth!: AuthenticationType;
+  private authHandlers!: Map<string, AuthenticationType>;
+
+  constructor(
+    config: SecurityPluginConfigType,
+    sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    router: IRouter,
+    esClient: ILegacyClusterClient,
+    coreSetup: CoreSetup,
+    logger: Logger
+  ) {
+    super(config, sessionStorageFactory, router, esClient, coreSetup, logger);
+    this.authTypes = this.config.auth.type;
+    this.authHandlers = new Map<string, AuthenticationType>();
+    this.init();
+  }
+
+  private async init() {
+    const routes = new MultiAuthRoutes(
+      this.router,
+      this.config,
+      this.sessionStorageFactory,
+      this.securityClient,
+      this.coreSetup
+    );
+    routes.setupRoutes();
+
+    for (let i = 0; i < this.authTypes.length; i++) {
+      switch (this.authTypes[i].toLowerCase()) {
+        case AuthType.BASIC: {
+          this.BasicAuth = new BasicAuthentication(
+            this.config,
+            this.sessionStorageFactory,
+            this.router,
+            this.esClient,
+            this.coreSetup,
+            this.logger
+          );
+          this.authHandlers.set(AuthType.BASIC, this.BasicAuth);
+          break;
+        }
+        case AuthType.OPEN_ID: {
+          this.OidcAuth = new OpenIdAuthentication(
+            this.config,
+            this.sessionStorageFactory,
+            this.router,
+            this.esClient,
+            this.coreSetup,
+            this.logger
+          );
+          this.authHandlers.set(AuthType.OPEN_ID, this.OidcAuth);
+          break;
+        }
+        case AuthType.SAML: {
+          this.SamlAuth = new SamlAuthentication(
+            this.config,
+            this.sessionStorageFactory,
+            this.router,
+            this.esClient,
+            this.coreSetup,
+            this.logger
+          );
+          this.authHandlers.set(AuthType.SAML, this.SamlAuth);
+          break;
+        }
+        default: {
+          throw new Error(`Unsupported authentication type: ${this.authTypes[i]}`);
+        }
+      }
+    }
+  }
+
+  // override functions inherited from AuthenticationType
+  async requestIncludesAuthInfo(
+    request: OpenSearchDashboardsRequest<unknown, unknown, unknown, any>
+  ): Promise<boolean> {
+    const cookie = await this.sessionStorageFactory.asScoped(request).get();
+    const reqAuthType = cookie?.authType?.toLowerCase();
+
+    if (reqAuthType && this.authHandlers.has(reqAuthType)) {
+      return this.authHandlers.get(reqAuthType)!.requestIncludesAuthInfo(request);
+    } else {
+      return false;
+    }
+  }
```

Comment:
I understand the `requestIncludesAuthInfo` is not a good naming here. The way OSD security works here is that, it tries to check if the request includes auth info in its header or URL. e.g. a REST client may use auth header to call OSD api directly, without using cookie. If there is no auth info like header in the request, then fall back to try to check if the cookie contains the auth info. 

You can refer to: https://github.com/aoguan1990/security-dashboards-plugin/blob/3ea9251238b420d656e90b7b721c1574f5f60767/server/auth/types/authentication_type.ts#L83-L87

Reply:
as mentioned, e.g. a `curl` cmd may call OSD API with `authorization` header in the request, without providing cookie, we need to handle that use case.

In this function, you can go over all configured auth type's `requestIncludesAuthInfo` and return true if one of them returns true

Reply:
Addressed.